File: Python/dynamics.py
import parameters as param
import numpy as np
from scipy import linalg as la
from scipy import constants as sc
import spin_matrices as sp
import functions as fn


def dynamics(microwave_amplitude):
    """ Solid effect dynamics, needs generalizing and optimising.
    """

    # Pre-allocate arrays for Hamiltonian and propagator
    hamiltonian = np.zeros((int(param.time_step_num), 2 ** param.num_spins, 2 ** param.num_spins,), dtype=np.complex)
    prop = np.zeros((int(param.time_step_num), 4 ** param.num_spins, 4 ** param.num_spins), dtype=np.complex)

    # Pre-allocate arrays for polarisation
    iznew = np.zeros(param.nrot)
    sznew = np.zeros(param.nrot)

    # Calculate time independent electron g-anisotropy coefficients
    c0, c1, c2, c3, c4 = fn.anisotropy_coefficients(param.orientation_tempol)

    # Calculate thermal density matrix
    density_mat = fn.density_mat_thermal()

    # Construct intrinsic Hamiltonian (Hilbert space)
    for count in range(0, int(param.time_step_num)):

        hyperfine_zz, hyperfine_zx = fn.hyperfine(count * param.time_step)

        ganisotropy = fn.anisotropy(c0, c1, c2, c3, c4, count * param.time_step)

        hamiltonian[count, :] = (ganisotropy - param.microwave_frequency) * sp.spin2_s_z + \
                                param.freq_nuclear_1 * sp.spin2_i_z + hyperfine_zz + hyperfine_zx

    # Calculate eigenvalues and eigenvectors of intrinsic Hamiltonian
    eigvals, eigvectors = np.linalg.eig(hamiltonian)
    energies = np.real(eigvals)
    eigvectors_inv = np.linalg.inv(eigvectors)

    # Calculate microwave Hamiltonian
    microwave_hamiltonian_0 = microwave_amplitude * sp.spin2_s_x

    # Calculate Liouville space propagator with relaxation
    for count in range(0, int(param.time_step_num)):

        microwave_hamiltonian = np.matmul(eigvectors_inv[count],
                                          np.matmul(microwave_hamiltonian_0, eigvectors[count]))

        # Calculate total Hamiltonian
        total_hamiltonian = np.diagflat(energies[count]) + microwave_hamiltonian

        # Transform spin matrices into time dependent basis
        Sxt, Syt, Szt, Ixt, Iyt, Izt, Spt, Smt, Ipt, Imt = fn.basis_transform(eigvectors, eigvectors_inv, count)

        # Transform Sz and Iz matrices (2^N to 4^N)
        LSzt = np.kron(Szt, np.transpose(Szt))
        LIzt = np.kron(Izt, np.transpose(Izt))

        # TODO Significance of these matrices?
        LIpt = np.kron(Ipt, np.transpose(Imt)) - 0.5 * np.eye(4 ** param.num_spins) + 0.5 * (
            np.kron(Izt, np.eye(2 ** param.num_spins)) + np.kron(np.eye(2 ** param.num_spins), np.transpose(Izt)))
        LImt = np.kron(Imt, np.transpose(Ipt)) - 0.5 * np.eye(4 ** param.num_spins) - 0.5 * (
            np.kron(Izt, np.eye(2 ** param.num_spins)) + np.kron(np.eye(2 ** param.num_spins), np.transpose(Izt)))
        LSpt = np.kron(Spt, np.transpose(Smt)) - 0.5 * np.eye(4 ** param.num_spins) + 0.5 * (
            np.kron(Szt, np.eye(2 ** param.num_spins)) + np.kron(np.eye(2 ** param.num_spins), np.transpose(Szt)))
        LSmt = np.kron(Smt, np.transpose(Spt)) - 0.5 * np.eye(4 ** param.num_spins) - 0.5 * (
            np.kron(Szt, np.eye(2 ** param.num_spins)) + np.kron(np.eye(2 ** param.num_spins), np.transpose(Szt)))

        # Calculate relaxation matrices
        Rt2ematv3 = (1 / param.t2_elec) * (LSzt - 0.25 * np.eye(4 ** param.num_spins))
        Rt2nmatv3 = (1 / param.t2_nuc) * (LIzt - 0.25 * np.eye(4 ** param.num_spins))
        Rt1mat = param.gep * LSpt + param.gem * LSmt + param.gnp * LIpt + param.gnm * LImt
        Rtot = Rt1mat + Rt2ematv3 + Rt2nmatv3

        # Calculate Liouville space Hamiltonian
        Lhamilt = np.kron(total_hamiltonian, np.eye(4)) - np.kron(np.eye(4), np.transpose(total_hamiltonian))

        # Calculate Liouville space eigenvectors
        LD = np.kron(eigvectors[count], eigvectors[count])
        LDinv = np.kron(eigvectors_inv[count], eigvectors_inv[count])

        # Calculate Liouville space propagator
        Ltot = Lhamilt + 1j * Rtot
        prop[count, :] = np.matmul(LDinv, np.matmul(la.expm(-1j * Ltot * param.time_step), LD))

    # TODO Significance of this loop
    prop_accu = np.eye(16)
    for count in range(0, int(param.time_step_num)):
        prop_accu = np.matmul(prop_accu, prop[count, :])

    for count in range(0, param.nrot):

        # Calculate electronic and nuclear polarisation
        iznew[count] = np.trace(np.matmul(np.real(density_mat), sp.spin2_i_z))
        sznew[count] = np.trace(np.matmul(np.real(density_mat), sp.spin2_s_z))

        # Transform density matrix (4x4 to 16x1)
        density_mat_liouville = np.reshape(density_mat, [16, 1])

        # Propagate density matrix
        density_mat = np.matmul(prop_accu, density_mat_liouville)

        # Transform density matrix (16x1 to 4x4)
        density_mat = np.reshape(density_mat, [4, 4])

    return iznew, sznew

File: Matlab/Chris/dynamics.m
function[iznew,sznew,iz_rot,sz_rot,evalgham]=dynamics(a,b,g,nr,final_time,mw_pwr,wme,t1e,t1n)

%%% Spin matrices (easyspin libriary)
spins = [1/2 1/2]; % S=1/2, I=1/2 system
nsp = length(spins); % Number of spins
Sz=sop(spins,'ze');
Iz=sop(spins,'ez');
Ix=sop(spins,'ex');
Iy=sop(spins,'ey');
IxSz=sop(spins,'zx');
IySz=sop(spins,'zy');
IzSz=sop(spins,'zz');
Sp=sop(spins,'+e');
Sm=sop(spins,'-e');
Ip=sop(spins,'e+');
Im=sop(spins,'e-');
Sx=sop(spins,'xe');
Sy=sop(spins,'ye');

%%% Parameters
wn=-400.9e6; % Nuclear frequency
we=28.025*9.4*1e9; % Electron frequency (B=9.4T)
w1=mw_pwr*1e6; % Convert microwave frequency to MHz
spinning_freq=nr; % Rotor frequency
wr = spinning_freq*1000; % Rotor frequency
tr = 1/wr; % Rotor period
nsteps = 10000; % Number of timesteps (absolute)
trstep = tr/nsteps; % Value of timestep (changes with rotor period)
tarray=0:trstep:tr; % Time array
hzz_max=3e6; % Hyperfine coupling
beta_en=0;
gamma_en=0;
hbar=1.054e-34; % Reduced planck
planck=hbar*2*pi; % planck
kb=1.3806e-23; % Boltzmann
T=100; % Temperature
t2e=1e-6; % T2 electron
t2n=1e-3; % T2 nucleus
pl=planck; % planck
h_mw=w1*Sx; % Microwave Hamiltonian
nrot=round(final_time/tr); % Number of data points?
iznew=zeros(1,nrot); % polarisation Iz
sznew=iznew; % polarisation Sz

% What are these, something to do with Liouvillian?
t_corr=pl/(kb*T);
p_e=tanh(0.5*we*t_corr);
p_n=tanh(0.5*wn*t_corr);
gnp=0.5*(1-p_n)*(1/(1*t1n)); % This is thermalisation, expansion of exp-(t_corr*H)/Trace(same);
gnm=0.5*(1+p_n)*(1/(1*t1n));
gep=0.5*(1-p_e)*(1/(1*t1e));
gem=0.5*(1+p_e)*(1/(1*t1e));

%%% G-anisotropy
alpha_g=253.6;
beta_g=105.1;
gamma_g=123.8;

gx=we*(2.00614/2);
gy=we*(2.00194/2);
gz=we*(2.00988/2);

ca = cosd(alpha_g);
cb = cosd(beta_g);
cg = cosd(gamma_g);
sa = sind(alpha_g);
sb = sind(beta_g);
sg = sind(gamma_g);

r11 = ca*cb*cg-sa*sg;
r12 = sa*cb*cg+ca*sg;
r13 = -sb*cg;
r21 = -ca*cb*sg-sa*cg;
r22 = -sa*cb*sg+ca*cg;
r23 = sb*sg;
r31 = ca*sb;
r32 = sa*sb;
r33 = cb;
c0 = 1/3*(gx+gy+gz);
c1 = 2*sqrt(2)/3*(gx*r11*r31+gy*r12*r32+gz*r13*r33);
c2 = 2*sqrt(2)/3*(gx*r21*r31+gy*r22*r32+gz*r23*r33);
c3 = 1/3*(gx*(r11^2-r21^2)+gy*(r12^2-r22^2)+gz*(r13^2-r23^2));
c4 = 2/3*(gx*r11*r21+gy*r22*r12+gz*r13*r23);

%%% Hamiltonian
hamil=zeros(2^nsp,2^nsp,nsteps);

for ii = 0:length(tarray)-2
    hhyp_zz=hzz_max*(-0.5*(sind(beta_en)^2)*cosd(2*(360*wr*ii*trstep+gamma_en))+...
        sqrt(2)*sind(beta_en)*cosd(beta_en)*cosd(360*wr*ii*trstep+gamma_en))*IzSz;
    hhyp_zx=hzz_max*(-0.5*sind(beta_en)*cosd(beta_en)*cosd(360*wr*ii*trstep+gamma_en)-...
        (sqrt(2)/4)*(sind(beta_en)^2)*cosd(2*(360*wr*ii*trstep+gamma_en))+(sqrt(2)/4)*(3*(cosd(beta_en)^2)-1))*IxSz;
    hhyp_zy=hzz_max*((sqrt(6)/4)*(sind(beta_en)^2)*sind(2*(360*wr*ii*trstep+gamma_en))-...
        (sqrt(3)/2)*sind(beta_en)*cosd(beta_en)*sind(1*(360*wr*ii*trstep+gamma_en)))*IySz;
    hhyp=2*(hhyp_zz+hhyp_zx+0*hhyp_zy);
    ganisohamil=c0+c1*cosd(360*wr*ii*trstep)+c2*sind(360*wr*ii*trstep)+c3*cosd(360*wr*ii*trstep*2)+c4*sind(360*wr*ii*trstep*2);
    hamil(:,:,ii+1) = (ganisohamil-wme)*Sz+wn*Iz+1*hhyp;
end

% Density matrix
Zp=exp((we+wn)*planck/(kb*T)) + exp((-we+wn)*planck/(kb*T))+exp((we-wn)*planck/(kb*T))+exp(-(we+wn)*planck/(kb*T));
% Presumably same as (1/np.sum(boltzmann_factors)) * np.diagflat(boltzmann_factors)
rho_zeeman=(1/Zp)*expm(-(we*Sz+wn*Iz)*planck/(kb*T));
rho0t=rho_zeeman;

% disp('rho0t')
% disp(rho0t)

% Below obselete?
ps0=trace(rho_zeeman*Sz);
pi0=trace(rho_zeeman*Iz);
p0=pi0+ps0;

% Calculate and sort eigenvalues
[evecham,evalgham]=eigenshuffle(hamil); % eigenvectors, eigenvalues

% disp('hamil')
% disp(size(hamil))
% disp(hamil(:, :, 1))

% disp('evecham')
% disp(size(evecham))
% disp(evecham(:, :, 1))
% 
% disp('evalgham')
% disp(size(evalgham))
% disp(evalgham(:,1))

%%% Propogation
prop=zeros(16,16,10000);

for ii=1:10000 % nsteps = 10000
%for ii=1:1 % nsteps = 10000
    D=squeeze(evecham(:,:,ii));
    Dinv=inv(D);
    hamil_mwt=Dinv*h_mw*D; % Transform the MW Hamiltonian to the same frame
%     disp('hamil_mwt')
%     disp(hamil_mwt)
    
    hamil_0=diag(evalgham(:,ii)); % Intrinsic Hamiltonian
    hamilt=hamil_0+hamil_mwt; % Total Hamiltonian (intrinsic + MW)
%     disp('hamilt')
%     disp(hamilt)
    
    % Tilt all the operators to the same frame
    Szt=Dinv*Sz*D;
    Izt=Dinv*Iz*D;
    Sxt=Dinv*Sx*D;
    Ixt=Dinv*Ix*D;
    Syt=Dinv*Sy*D;
    Iyt=Dinv*Iy*D;
    Spt=Dinv*Sp*D;
    Smt=Dinv*Sm*D;
    Ipt=Dinv*Ip*D;
    Imt=Dinv*Im*D;
%     
%     disp('Szt')
%     disp(Szt)
%     
%     disp('Im')
%     disp(Im)
%     
%     disp('Imt')
%     disp(Imt)
%     
%     disp('Sp')
%     disp(Sp)
%     
%     disp('Spt')
%     disp(Spt)
    
    % Liouvillian
    LSzt=kron(Szt,Szt.');
    
%     disp('LSzt')
%     disp(LSzt)
  
    RSzt=kron(eye(2^nsp),eye(2^nsp));
    LIzt=kron(Izt,Izt.');
    RIzt=kron(eye(2^nsp),eye(2^nsp)); % equal to RSzt
    
    LIpt=1.0*kron(Ipt,Imt.')-.5*eye(4^nsp)+.5*(kron(Izt,eye(2^nsp))+kron(eye(2^nsp),Izt.'));
%     disp('LIpt')
%     disp(LIpt)
    
    LImt=1.0*kron(Imt,Ipt.')-.5*eye(4^nsp)-.5*(kron(Izt,eye(2^nsp))+kron(eye(2^nsp),Izt.'));
    LSpt=1.0*kron(Spt,Smt.')-.5*eye(4^nsp)+.5*(kron(Szt,eye(2^nsp))+kron(eye(2^nsp),Szt.'));
    LSmt=1.0*kron(Smt,Spt.')-.5*eye(4^nsp)-.5*(kron(Szt,eye(2^nsp))+kron(eye(2^nsp),Szt.'));
    
    Rt2ematv3=(1/t2e)*(LSzt-.25*RSzt);
    Rt2nmatv3=(1/(t2n))*(LIzt-.25*RIzt);
    Rt1mat=gep*LSpt+gem*LSmt+gnp*LIpt+gnm*LImt;
    Rtot=1*Rt1mat+1*Rt2ematv3+1*Rt2nmatv3;
    
    Lhamilt=kron((hamilt),eye(4))-kron(eye(4),(hamilt).');
%     disp('Lhamilt')
%     disp(Lhamilt)
    
    LD=kron(D,D);
    LDinv=kron(Dinv,Dinv);
    Ltot=Lhamilt+1*1i*((Rtot));
    test = LD*expm(-1i*Ltot*trstep)*LDinv;
    prop(:,:,ii)=LD*expm(-1i*Ltot*trstep)*LDinv; % Transformation back into Zeeman frame?
    
%     disp('test')
%     disp(test)
end

prop_accu=eye(16); % 16x16 identity matrix
for kk=1:nsteps % Why multiply identity matrix with prop?
    prop_accu=prop_accu*squeeze(prop(:,:,kk));
end

% Calculate polarisation? We have 10000 timesteps, so what is nrot?
for jj=1:nrot
    iznew(jj)=trace(real(rho0t)*Iz);
    sznew(jj)=trace(real(rho0t)*Sz);
    Lrho0t=reshape(rho0t,[16,1]);
    rho0t=prop_accu*Lrho0t;
    rho0t=reshape(rho0t,[4,4]);
end

rho0t=rho_zeeman;

for ii=1:nsteps
        iz_rot(ii)=trace(rho0t*Iz); 
        sz_rot(ii)=trace(rho0t*Sz);
        Lrho0t=reshape(rho0t,[16,1]);
        rho0t=squeeze(prop(:,:,ii))*Lrho0t;
        rho0t=reshape(rho0t,[4,4]); 
end

end